src/utils/OnboardingProvider.tsx

```diff
@@ -0,0 +1,293 @@
+  setProgress: Dispatch<SetStateAction<number>>;
+  setCanContinue: Dispatch<SetStateAction<boolean>>;
+  setFlow: Dispatch<SetStateAction<FlowData[]>>;
+  setLanguages: (languages: ProfileLanguage[]) => Promise<void>;
+  setRoles: (roles: ProfileRole[]) => Promise<void>;
+}
+
+export const OnboardingContext = createContext<
+  OnboardingContextType | undefined
+>(undefined);
+
+const blankProfile: Profile = {
+  first_name: '',
+  last_name: '',
+  hours_per_month: 0,
+  location: '',
+  start_date: '',
+  user_id: '0-0-0-0-0',
+};
+
+export default function OnboardingProvider({
+  children,
+}: {
+  children: ReactNode;
+}) {
+  const [progress, setProgress] = useState(0);
+  const [flow, setFlow] = useState<FlowData[]>([]);
+  const [profile, setProfile] = useState<Profile>({ ...blankProfile });
+  const [languages, setProfileLangs] = useState<ProfileLanguage[]>([]);
+  const [roles, setProfileRoles] = useState<ProfileRole[]>([]);
+  const [canContinue, setCanContinue] = useState<boolean>(false);
+
+  useEffect(() => {
+    (async () => {
+      const { data: sessionData, error } = await supabase.auth.getSession();
+
+      if (error) throw error;
+
+      const userId = sessionData?.session?.user.id;
+      if (userId) {
+        setProfile({ ...blankProfile, user_id: userId as UUID });
+      }
+    })();
+  }, []);
+
+  const providerValue = useMemo(() => {
+    /**
+     * Updates stored profile state with the partial data.
+     * Does not affect database.
+     */
+    const updateProfile = async (updatedInfo: Partial<Profile>) => {
+      const newProfileData = { ...profile, ...updatedInfo };
+      setProfile(newProfileData);
+    };
+
+    /**
+     * Flushes stored data with Supabase.
+     * Requires RLS insert and update policy on the profiles table.
+     */
+    const flushProfileData = async () => {
+      // validate profile
+      if (profile.first_name === blankProfile.first_name)
+        throw new Error('Profile is missing first name!');
+
+      if (profile.last_name === blankProfile.last_name)
+        throw new Error('Profile is missing last name!');
+
+      if (profile.hours_per_month === blankProfile.hours_per_month)
+        throw new Error('Profile is missing hours per month!');
+
+      if (profile.location === blankProfile.location)
+        throw new Error('Profile is missing location!');
+
+      if (profile.start_date === blankProfile.start_date)
+        throw new Error('Profile is missing start date!');
+
+      if (profile.user_id === blankProfile.user_id)
+        throw new Error('Profile is missing user id!');
+
+      // upsert
+      await upsertProfile(profile);
+    };
+
+    /**
+     * Takes in a list of ProfileLanguage
+     * and replaces languages with it entirely,
+     * simultaneously updating the database.
+     * Requires RLS delete, update, insert on profiles-languages.
+     */
+    const setLanguages = async (newLangs: ProfileLanguage[]) => {
```

Comment:
clarification: are we setting the languages and roles on supabase via the onboarding context, but not the profile? ifl we should be consistent and not allow the onboarding context to write to supabase at all, and ultimately use the ProfileContext at the end to actually make changes on supabase - esp because they may go back and change their selections for roles and languages

Reply:
yee true, will update